src/openrct2/world/Map.cpp

```diff
@@ -1539,6 +1548,104 @@ void map_update_tiles()
     }
 }
 
+void MapInvalidateHeightCache()
+{
+    _mapHighestTileHeightLoopPosition = 0;
+}
+
+void MapUpdateHeightCache(bool fullCheck)
+{
+    constexpr int32_t mapTileCount = MAXIMUM_MAP_SIZE_TECHNICAL * MAXIMUM_MAP_SIZE_TECHNICAL;
+
+    if (fullCheck)
+        _mapHighestTileHeightLoopPosition = 0;
+
+    if (_mapHighestTileHeightLoopPosition == mapTileCount - 1)
+        return;
+
+    // If we're doing a full check of the map then zero out the loop
+    // position to start at tile zero, checking all tiles before returning.
+    // Otherwise check a limited number of tiles per each time the
+    // function is called.
+    int32_t tilesToCheck = fullCheck ? mapTileCount : mapTileCount / 64;
+
+    for (int32_t i = 0; i < tilesToCheck; i++)
+    {
+        int32_t x = (_mapHighestTileHeightLoopPosition % MAXIMUM_MAP_SIZE_TECHNICAL) * COORDS_XY_STEP;
+        int32_t y = (_mapHighestTileHeightLoopPosition / MAXIMUM_MAP_SIZE_TECHNICAL) * COORDS_XY_STEP;
+        auto tileCoords = CoordsXY(x, y);
+        auto tileElement = map_get_first_element_at(tileCoords);
+
+        do
+        {
+            if (tileElement == nullptr)
+                break;
+
+            if (tileElement->GetType() == TILE_ELEMENT_TYPE_CORRUPT)
+                continue;
+
+            // Here we are handling Water height, which requires a bit of different approach.
+            if (tileElement->GetType() == TILE_ELEMENT_TYPE_SURFACE)
+            {
+                int32_t waterHeight = tileElement->AsSurface()->GetWaterHeight();
+                if (waterHeight > _mapHighestTileHeightPending)
+                {
+                    _mapHighestTileHeightPending = waterHeight;
+                    continue;
+                }
+            }
+
+            int32_t height = tileElement->GetClearanceZ();
+            if (height > _mapHighestTileHeightPending)
+                _mapHighestTileHeightPending = height;
+
+        } while (!(tileElement++)->IsLastForTile());
+
+        if (_mapHighestTileHeightLoopPosition == mapTileCount - 1)
+        {
+            _mapHighestTileHeight = _mapHighestTileHeightPending;
+            _mapHighestTileHeightPending = 0;
+            break;
+        }
+        else
+            _mapHighestTileHeightLoopPosition++;
```

Comment:
If you move this above the ```if``` then you can remove ```- 1``` in the test and at the top of the function.

Reply:
Sounds good, done 👍